state/relation.go

```diff
@@ -149,6 +152,100 @@ func (r *Relation) SetStatus(statusInfo status.StatusInfo) error {
 	})
 }
 
+// SetSuspended sets whether the relation is suspended.
+func (r *Relation) SetSuspended(suspended bool) error {
+	var buildTxn jujutxn.TransactionSource = func(attempt int) ([]txn.Op, error) {
+
+		if attempt > 1 {
+			if err := r.Refresh(); err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+
+		// Relation must be associated with an offer.
+		// TODO(wallyworld) - ultimately this will not be the case when we can resume a relation
+		// on the consuming side. We need to delete macaroons when a relation is suspended though
+		// before we can do that.
+		oc, err := r.st.OfferConnectionForRelation(r.Tag().Id())
+		if errors.IsNotFound(err) {
+			return nil, errors.Errorf("cannot set status for %q which is not associated with an offer", r.Tag().Id())
+		}
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		checkOps := []txn.Op{{
+			C:      offerConnectionsC,
+			Id:     fmt.Sprintf("%d", r.Id()),
+			Assert: txn.DocExists,
+		}}
+
+		currentStatus, err := r.Status()
```

Comment:
You shouldn't care about the status anymore, just the suspended flag. And I don't think it really matters if it's already suspended does it? Just return nil if suspended==doc.Suspended?

Reply:
The intent was not to allow a relation to be suspended if it is in error, only if it is running normally.
etc

Reply:
I don't see a compelling reason to prevent joining/suspending based on it being in an error status. Business rules should be based on the model state, not current operational status.

What's the use case for preventing suspend when it's in an error status? IIUC, it'll only go to error status if there's an issue with networking. Why should that prevent joining or suspending?

Reply:
If it's in error, it's not expected to be able to, say, resume again until the source of the error is fixed, eg firewall rules adjusted. I guess we can let the resume happen and the relation status will then reflect the outcome of that.